src/main/java/net/minecraftforge/registries/DeferredRegister.java

```diff
@@ -99,24 +156,111 @@ public void register(IEventBus bus)
     }
 
     /**
+     * Since the view is unmodifiable, can't cache it really. Gotta stream and collect each time
+     * This should not be a costly operation however, since the {@link RegistryObject} is cached.
+     * The modders can also cache the entries themselves.
      * @return The unmodifiable view of registered entries. Useful for bulk operations on all values.
      */
     public Collection<RegistryObject<T>> getEntries()
     {
+        if(entriesView.isEmpty())
+        {
+            return lazyEntriesView.stream().map(LazyRegistryObject::getRegistryObject).collect(Collectors.toSet());
+        }
         return entriesView;
     }
 
     private void addEntries(RegistryEvent.Register<?> event)
     {
-        if (event.getGenericType() == this.type.getRegistrySuperType())
+        if (event.getGenericType() == this.getFromSupplier().getRegistrySuperType())
         {
             @SuppressWarnings("unchecked")
             IForgeRegistry<T> reg = (IForgeRegistry<T>)event.getRegistry();
-            for (Entry<RegistryObject<T>, Supplier<? extends T>> e : entries.entrySet())
+            if(!entries.isEmpty())
+            {
+                for (Entry<RegistryObject<T>, Supplier<? extends T>> e : entries.entrySet())
+                {
+                    reg.register(e.getValue().get());
+                    e.getKey().updateReference(reg);
+                }
+            }
+            else
             {
-                reg.register(e.getValue().get());
-                e.getKey().updateReference(reg);
+                for (Entry<LazyRegistryObject<T>, Supplier<? extends T>> e : lazyEntries.entrySet())
+                {
+                    reg.register(e.getValue().get());
+                    e.getKey().getRegistryObject().updateReference(reg);
+                }
             }
         }
     }
+
+    /**
+     * If the IForgeRegistry is null, gets it from the supplier. This is sure to be called from a classic register event.
+     * It is even guaranteed to be called from the Register Block event, since it is the first one called.
+     * This also makes sure that the IForgeRegistry is correct when updating the RegistryObjects.
+     *
+     * This is only useful for custom registries, when the {@link IForgeRegistry} does not exist when static init happens
+     */
+    private IForgeRegistry<T> getFromSupplier(){
+        if(type == null)
+        {
+            type = typeSup.get();
+        }
+        return type;
+    }
+
+    /**
+     * A way of lazily evaluating registry objects but in a nicer way than using {@link LazyOptional<RegistryObject>}
+     *
+     * Also a supplier, to mimic {@link RegistryObject} as best as possible
+     */
+    public static class LazyRegistryObject<T extends IForgeRegistryEntry<? super T>> implements Supplier<T>
+    {
+        private final LazyOptional<RegistryObject<T>> lazyRet;
```

Comment:
Instead of using `LazyOptional` which I feel like is being misused here, this should use [`net.minecraftforge.common.util.NonNullLazy`](https://github.com/MinecraftForge/MinecraftForge/blob/1.15.x/src/main/java/net/minecraftforge/common/util/NonNullLazy.java).

Reply:
Seems indeed better. hesitated to use Lazy directly for some reason, will use forges versions